### PR TITLE
fix: failed to execute sql with subquery

### LINKED ISSUE #5529: Failed to execute sql with subquery
**Describe the bug**

Datafusion failed to execute the sql that has its subquery wrapped in two or more parentheses. See the reproduce.

**To Reproduce**

In datafusion-cli:

```text
DataFusion CLI v19.0.0
❯ CREATE EXTERNAL TABLE foo (a INT, b INT) STORED AS CSV LOCATION 'data.csv';
0 rows in set. Query took 0.009 seconds.
❯ SELECT * FROM foo WHERE a IN ((SELECT a FROM foo));
This feature is not implemented: Physical plan does not support logical expression (<subquery>)
❯ explain SELECT * FROM foo WHERE a IN ((SELECT a FROM foo));
+--------------+----------------------------------------------------------------------------+
| plan_type    | plan                                                                       |
+--------------+----------------------------------------------------------------------------+
| logical_plan | Filter: foo.a = (<subquery>)                                               |
|              |   Subquery:                                                                |
|              |     Projection: foo.a                                                      |
|              |       TableScan: foo                                                       |
|              |   TableScan: foo projection=[a, b], partial_filters=[foo.a = (<subquery>)] |
|              |     Subquery:                                                              |
|              |       Projection: foo.a                                                    |
|              |         TableScan: foo                                                     |
+--------------+----------------------------------------------------------------------------+
1 row in set. Query took 0.008 seconds.
```

If the subquery is wrapped in "(( ))" instead of "( )", datafusion failed with message "`This feature is not implemented: Physical plan does not support logical expression (<subquery>)`".

**Expected behavior**

Both mysql and postgresql can execute that sql. And when subquery is wrapped in single pair of parenthesis, the sql got executed correctly in datafusion, too.

**Additional context**

I'm a little hesitate to call it a bug. Clearly the subquery logical plan node is not supported in physical plan, yet. However, after some investigation on the codes, I think we can come up with a walkaround in the mean time.

It's because the subquery "`where a in (( select a from foo ))`" got interpreted to `Expr::InList`, while the one with "( )" is `Expr::InSubquery`. Datafusion optimizes the `Expr::InSubquery` in `DecorrelateWhereIn` to join, effectively not producing any subqueries that physical plan cannot run.

So the problem could be solved if we can make this specific "(( ))" case interpreted as `Expr::InSubquery`. That might be easy to be done in `ExprRewriter::mutate`:

https://github.com/apache/arrow-datafusion/blob/ff013e24559850541c54cf5d254922649ef37d8c/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs#L395-L399

WDYT? @alamb @waynexia

### PR DESCRIPTION
**Pull request by MichaelScofield**
# Which issue does this PR close?

Closes #5529 .

# Rationale for this change

fix: failed to run "where x in ((select ...))"

# What changes are included in this PR?

rewrite `Expr::InList` with only one subquery to `Expr::InSubquery` during logical plan optimization

# Are these changes tested?

yes

# Are there any user-facing changes?

no

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 1]
```diff
@@ -390,6 +390,22 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                 lit(negated)
             }
 
+            // expr IN ((subquery)) -> expr IN (subquery), see ##5529
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:393]: 👍

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 2]
```diff
@@ -2849,6 +2866,24 @@ mod tests {
             simplify(in_list(col("c1"), vec![lit(1), lit(2)], true)),
             col("c1").not_eq(lit(2)).and(col("c1").not_eq(lit(1)))
         );
+
+        let subquery = Arc::new(test_table_scan_with_name("test").unwrap());
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:2870]: Can you please also add a negative test (with two subqueries)?

**Review by alamb — COMMENTED** [review]
Thank you @MichaelScofield !

This looks great to me.

Can you please add one SQL level test to verify that the plumbing is all hooked up correctly.

I recommend using in https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/sqllogictests 

Perhaps in https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sqllogictests/test_files/subquery.slt ?

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 3]
```diff
@@ -390,6 +390,22 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
 
+            // expr IN ((subquery)) -> expr IN (subquery), see ##5529
+            Expr::InList {
+                expr,
+                mut list,
+                negated,
+            } if list.len() == 1
+                && matches!(list.first(), Some(Expr::ScalarSubquery { .. })) =>
+            {
+                let Expr::ScalarSubquery(subquery) = list.remove(0) else { unreachable!() };
```
Comment (comphead) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:401]: Struggling to get why we unreachable here

Comment (waynexia) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:401]: The condition in L399 ensures that only `ScalarSubquery` is allowed, making other branches unreachable. Although there is an unstable [rust feature](https://github.com/rust-lang/rust/issues/51114) available to simplify this, it cannot be used at the moment.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs (hunk 2 above)
Comment (MichaelScofield) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:2870]: of course

**Comment by MichaelScofield** [comment]
I've added 3 test cases in sqllogictest. The cases are from our original intention to test subquery with "(( ))". PTAL @alamb

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @MichaelScofield

**Pull request merged**